IPython/extensions/autoreload.py

```diff
@@ -743,6 +872,36 @@ def post_execute_hook(self):
 
         self.loaded_modules.update(newly_loaded_modules)
 
+    def _track_imports_from_code(self, code: str) -> None:
+        """Track import statements from executed code"""
+        try:
+            tree = ast.parse(code)
+
+            for node in ast.walk(tree):
+                # Handle "from X import Y" style imports
+                if isinstance(node, ast.ImportFrom):
+                    mod = node.module
+
+                    # Skip relative imports that don't have a module name
+                    if mod is None:
+                        continue
+
+                    for name in node.names:
+                        # name.name is going to be actual name that we want to import from module
+                        # name.asname is Z in the case of from X import Y as Z
+                        # we should update Z in the shell in this situation, so track it too.
+                        original_name = name.name
+                        resolved_name = name.asname if name.asname else name.name
+
+                        # Since the code executed successfully, we know this import is valid
+                        self._reloader.import_from_tracker.add_import(
+                            mod, original_name, resolved_name
+                        )
```

Comment:
This can track more imports than actually executed. Consider:

```
from a import b
if TYPE_CHECKING:
    from c import d
```

It will track both even if `TYPE_CHECKING` is false. I think this does not matter that much, but just noting it.

Reply:
Well, no it does still lead to unintended behaviour:

Start with:

![image](https://github.com/user-attachments/assets/ea1a03e9-b77a-4399-8bd7-3558c7a7a0ab)

and then modify `a.py` (still good so far):

![image](https://github.com/user-attachments/assets/60beb4c8-2e13-4d90-b88f-159ca43f5351)

modify `a.py` one more time, now change the `b` value:

![image](https://github.com/user-attachments/assets/ff01c47b-bd86-462a-9e16-9037bb1c306b)